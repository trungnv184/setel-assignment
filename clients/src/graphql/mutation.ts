import { gql } from '@apollo/client';

const ADD_PRODUCT_TO_CART = gql`
  mutation AddToCart($cartInput: CartInput) {
    addToCart(cartInput: $cartInput) {
      id
      name
      url
      price
    }
  }
`;

const CREATE_ORDER_PRODUCTS = gql`
  mutation CreateOrder($orderInput: OrderInput) {
    createOrder(orderInput: $orderInput) {
      orderItems {
        id
        name
        url
        price
        quantity
        createdDate
      }
      metadata {
        firstName
        lastName
        address
        phoneNumber
      }
      notes
    }
  }
`;

export { ADD_PRODUCT_TO_CART, CREATE_ORDER_PRODUCTS };
