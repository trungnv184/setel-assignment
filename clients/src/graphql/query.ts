import { gql } from '@apollo/client';

const GET_PRODUCTS = gql`
  query GetProducts {
    products {
      id
      name
      price
      url
      rating
    }
  }
`;

const GET_PRODUCTS_IN_CART = gql`
  query GetProductsCart {
    getProductsCart {
      id
      name
      price
      url
      quantity
    }
  }
`;

const GET_ORDER_BY_ID = gql`
  query GetOrder($orderId: String) {
    getOrder(orderId: $orderId) {
      id
      orderItems {
        id
        name
        price
        url
        quantity
      }
      metadata {
        firstName
        lastName
        address
        phoneNumber
      }
      notes
      state
      createdDate
      updatedDate
    }
  }
`;

const GET_ORDER_STATUS = gql`
  query GetOrderStatus($orderId: String) {
    getOrder(orderId: $orderId) {
      state
    }
  }
`;

export { GET_PRODUCTS, GET_PRODUCTS_IN_CART, GET_ORDER_BY_ID, GET_ORDER_STATUS };
