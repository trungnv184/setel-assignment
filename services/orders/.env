DELAY_DELIVERY = 60000
MONGO_URI = mongodb+srv://setel:3JQAQhzkjLuvjE9b@cluster0.9w2cg.mongodb.net/OrderPayments?retryWrites=true&w=majority