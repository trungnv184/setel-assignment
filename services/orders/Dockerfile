FROM node:12

WORKDIR /usr/src/app

COPY package*.json ./

RUN npm install

COPY . .

ENV MONGO_URI=mongodb+srv://setel:3JQAQhzkjLuvjE9b@cluster0.9w2cg.mongodb.net/OrderPayments?retryWrites=true&w=majority

ENV DELAY_DELIVERY=60000

CMD ["npm", "run", "start"]
