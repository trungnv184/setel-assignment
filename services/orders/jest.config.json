{
  "moduleDirectories": ["node_modules", "src"],
  "moduleFileExtensions": ["js", "json", "ts"],
  "testEnvironment": "node",
  "testRegex": ".spec.ts$",
  "transform": {
    "^.+\\.ts$": "ts-jest"
  },
  "coverageDirectory": "./coverage",
  "collectCoverage": true,
  "moduleNameMapper": {
    "^@schemas$": "<rootDir>/src/orders/schemas",
    "^@dto(.*)$": "<rootDir>/src/orders/dto$1",
    "^@configs(.*)$": "<rootDir>/src/orders/config$1",
    "^@common(.*)$": "<rootDir>/src/orders/common$1",
    "^@orders(.*)$": "<rootDir>/src/orders/$1",
    "^@utils(.*)$": "<rootDir>/src/orders/utils$1",
    "^@payments(.*)$": "<rootDir>/src/payments$1"
  }
}
