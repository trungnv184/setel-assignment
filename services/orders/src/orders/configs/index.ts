export const WAIT_FOR_PAYMENT_PROCESSING = 5000;
export const MONGO_URI =
  'mongodb+srv://setel:3JQAQhzkjLuvjE9b@cluster0.9w2cg.mongodb.net/OrderPayments?retryWrites=true&w=majority';
