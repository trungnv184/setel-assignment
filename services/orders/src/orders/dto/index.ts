export { CreateOrderDto } from './created-order.dto';
export { UpdatedOrderDto } from './updated-order.dto';
