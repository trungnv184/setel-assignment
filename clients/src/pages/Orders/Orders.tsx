import React from 'react';
import { ErrorMessage, OrderItems } from 'components';
import { useQuery } from '@apollo/client';
import { GET_ORDER_LIST } from 'graphql/query';
import { Dimmer, Loader, Message } from 'semantic-ui-react';
import { OrderResponse } from 'types/order-response';
import { StyledContainer } from 'components/Common/Style';
import { useHistory } from 'react-router-dom';

const Orders: React.FC<any> = () => {
  const { data, loading, error } = useQuery(GET_ORDER_LIST, {
    fetchPolicy: 'cache-and-network',
  });
  const history = useHistory();

  if (loading) {
    return (
      <Dimmer active={true}>
        <Loader />
      </Dimmer>
    );
  }

  if (error) return <ErrorMessage />;

  const orders = data?.getOrderList as OrderResponse[];

  const renderOrderList = () => {
    return orders.map((order) => (
      <StyledContainer key={order.id} onClick={() => navigateToOrderDetailPage(order.id)}>
        <OrderItems orderItems={order.orderItems} orderState={order.state} createdDate={order.createdDate} />
      </StyledContainer>
    ));
  };

  const navigateToOrderDetailPage = (orderId: string) => {
    history.push(`/order-detail/${orderId}`);
  };
  return (
    <>
      {orders.length > 0 ? (
        renderOrderList()
      ) : (
        <Message
          header="Your order history"
          content="You don't have any orders. Do you want to purchase some books today ?"
        />
      )}
    </>
  );
};

export default Orders;
