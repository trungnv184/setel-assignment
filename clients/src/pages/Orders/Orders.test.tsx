import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { GET_ORDER_LIST } from 'graphql/query';
import Orders from './Orders';
import { ORDERS_RESPONSE_MOCK } from 'test/mocks/orders.mock';

let isCalled = false;

const getMocks = (isEmptyData: boolean = false) => {
  const response = isEmptyData
    ? {
        data: {
          getOrderList: [],
        },
      }
    : ORDERS_RESPONSE_MOCK;
  return [
    {
      request: {
        query: GET_ORDER_LIST,
      },
      result: () => {
        isCalled = true;
        return response;
      },
    },
  ];
};

beforeEach(() => {
  jest.resetAllMocks();
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('OrdersPage', () => {
  test('Should render products page', async () => {
    const { container } = render(
      <MockedProvider mocks={getMocks()} addTypename={false}>
        <Orders />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(isCalled).toBeTruthy();
    });

    const bookNameElements = screen.getAllByTestId('book-name');

    // Total 5 books for 2 orders
    expect(bookNameElements.length).toBe(5);
    expect(bookNameElements[0].textContent).toBe('IT Book');
    expect(container).toMatchSnapshot();
  });

  test('Should display empty order message', async () => {
    render(
      <MockedProvider mocks={getMocks(true)} addTypename={false}>
        <Orders />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(isCalled).toBeTruthy();
    });

    expect(
      screen.getByText(/You don't have any orders. Do you want to purchase some books today ?/i)
    ).toBeInTheDocument();
  });
});
