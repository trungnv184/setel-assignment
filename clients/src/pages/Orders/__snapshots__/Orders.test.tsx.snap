// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersPage Should render products page 1`] = `
<div>
  <div
    class="ui container sc-gKAaRy OrlUr"
  >
    <div
      class="ui segment"
    >
      <div
        class="ui green label"
      >
        <i
          aria-hidden="true"
          class="book icon"
        />
         
        Selected Books 1
      </div>
      <div
        class="ui blue label"
      >
        <i
          aria-hidden="true"
          class="payment icon"
        />
         
        500 ₫
      </div>
      <div
        class="ui purple label"
      >
        <i
          aria-hidden="true"
          class="calendar icon"
        />
         
        20-06-2021
      </div>
      <div
        class="ui red label"
      >
        <i
          aria-hidden="true"
          class="truck icon"
        />
         
        Delivered
      </div>
      <div
        class="ui six column grid"
      >
        <div
          class="column"
        >
          <div
            class="sc-gtsrHT igIUx"
          >
            <img
              class="ui small image"
              src="url"
            />
            <div
              class="sc-dlnjwi guJDWs"
              data-testid="book-name"
            >
              IT Book
            </div>
            <div
              class="sc-dlnjwi guJDWs"
            >
              Price:
              500 ₫
            </div>
            <div
              class="sc-dlnjwi dHdVQU"
            >
              Quantity:
              1
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ui container sc-gKAaRy OrlUr"
  >
    <div
      class="ui segment"
    >
      <div
        class="ui green label"
      >
        <i
          aria-hidden="true"
          class="book icon"
        />
         
        Selected Books 4
      </div>
      <div
        class="ui blue label"
      >
        <i
          aria-hidden="true"
          class="payment icon"
        />
         
        800.000 ₫
      </div>
      <div
        class="ui purple label"
      >
        <i
          aria-hidden="true"
          class="calendar icon"
        />
         
        20-06-2021
      </div>
      <div
        class="ui red label"
      >
        <i
          aria-hidden="true"
          class="truck icon"
        />
         
        Delivered
      </div>
      <div
        class="ui six column grid"
      >
        <div
          class="column"
        >
          <div
            class="sc-gtsrHT igIUx"
          >
            <img
              class="ui small image"
              src="https://salt.tikicdn.com/cache/280x280/ts/product/5c/07/92/7ac1298303440baa2ed74fa20c9f7385.jpg"
            />
            <div
              class="sc-dlnjwi guJDWs"
              data-testid="book-name"
            >
              Làm chủ cảm xúc
            </div>
            <div
              class="sc-dlnjwi guJDWs"
            >
              Price:
              200.000 ₫
            </div>
            <div
              class="sc-dlnjwi dHdVQU"
            >
              Quantity:
              1
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="sc-gtsrHT igIUx"
          >
            <img
              class="ui small image"
              src="https://salt.tikicdn.com/cache/280x280/ts/product/59/e3/e7/642ce38017ac1d35f2e87234b6fd56f2.jpg"
            />
            <div
              class="sc-dlnjwi guJDWs"
              data-testid="book-name"
            >
              Tâm lý học hạnh phúc
            </div>
            <div
              class="sc-dlnjwi guJDWs"
            >
              Price:
              200.000 ₫
            </div>
            <div
              class="sc-dlnjwi dHdVQU"
            >
              Quantity:
              1
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="sc-gtsrHT igIUx"
          >
            <img
              class="ui small image"
              src="https://salt.tikicdn.com/cache/280x280/ts/product/5f/36/dd/c64c91016b320ae4591efcf93d0ebf30.jpg"
            />
            <div
              class="sc-dlnjwi guJDWs"
              data-testid="book-name"
            >
              Dạy con học ít hiểu nhiều
            </div>
            <div
              class="sc-dlnjwi guJDWs"
            >
              Price:
              200.000 ₫
            </div>
            <div
              class="sc-dlnjwi dHdVQU"
            >
              Quantity:
              1
            </div>
          </div>
        </div>
        <div
          class="column"
        >
          <div
            class="sc-gtsrHT igIUx"
          >
            <img
              class="ui small image"
              src="https://salt.tikicdn.com/cache/280x280/ts/product/95/c5/71/f260011660001af960c42d7d2f1c5dc1.jpg"
            />
            <div
              class="sc-dlnjwi guJDWs"
              data-testid="book-name"
            >
              Đừng chạy theo số đông
            </div>
            <div
              class="sc-dlnjwi guJDWs"
            >
              Price:
              200.000 ₫
            </div>
            <div
              class="sc-dlnjwi dHdVQU"
            >
              Quantity:
              1
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
